tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,744 @@
+            // do so until `added` is drained, so we don't need to block on the
+            // lock.
+            return;
+        }
+
+        // Otherwise, no other thread is assigning permits. Thus, we must lock
+        // the semaphore's wait list and assign permits until `added` is
+        // drained.
+        self.add_permits_locked(added, self.waiters.lock().unwrap(), false);
+    }
+
+    /// Closes the semaphore. This prevents the semaphore from issuing new
+    /// permits and notifies all pending waiters.
+    pub(crate) fn close(&self) {
+        // Closing the semaphore works similarly to adding permits: if another
+        // thread is already holding the lock on the wait list to assign
+        // permits, we simply set a bit in the counter of permits being added.
+        // The thread holding the lock will see that bit has been set, and begin
+        // closing the semaphore.
+
+        self.permits.fetch_or(CLOSED, Release);
+        // Acquire the `added`, setting the "closed" flag on the lock.
+        let prev = self.adding.fetch_or(1, AcqRel);
+
+        if prev != 0 {
+            // Another thread has the lock and will be responsible for notifying
+            // pending waiters.
+            return;
+        }
+
+        let mut lock = self.waiters.lock().unwrap();
+        lock.closed = true;
+        self.add_permits_locked(0, lock, true);
+    }
+
+    fn add_permits_locked(
+        &self,
+        mut rem: usize,
+        mut waiters: MutexGuard<'_, Waitlist>,
+        mut closed: bool,
+    ) {
+        // The thread adding permits will loop until the counter of added
+        // permits is drained. If threads add permits (or close the semaphore)
+        // while we are holding the lock, we will add those permits as well, so
+        // that the other thread does not need to block.
+
+        loop {
+            // How many permits are we releasing on this iteration?
+            let initial = rem;
+            // Assign permits to the wait queue and notify any satisfied
+            // waiters.
+            while rem > 0 || waiters.closed {
+                let pop = match waiters.queue.last() {
+                    Some(_) if waiters.closed => true,
+                    Some(last) => last.assign_permits(&mut rem, waiters.closed),
+                    None => {
+                        self.permits.fetch_add(rem, AcqRel);
+                        break;
+                    }
+                };
+                if pop {
+                    let node = waiters.queue.pop_back().unwrap();
+                    // Safety: we are holding the lock on the wait queue, and
+                    // thus have exclusive access to the nodes' wakers.
+                    let waker = unsafe { 
+                        node.as_ref().waker.with_mut(|waker| (*waker).take())
+                    };
+
+                    waker.expect("if a node is in the wait list, it must have a waker").wake();
```

Comment:
Waking should be done **outside** of the mutex. Otherwise, the woken task could run and hit a locked mutex.